File: algorithms/leet.0059.src.1.java
public class Solution {
    public int[][] generateMatrix(int n) {
        int[][] ans = new int[n][n];
        for(int[] row: ans) Arrays.fill(row, -1);
        int[] dx = { 0, 1, 0, -1 };
        int[] dy = { 1, 0, -1, 0 };
        int d = 0, x = 0, y = 0;
        for(int i = 0; i < n * n; ++i) {
            ans[x][y] = i + 1;
            if( 0 <= x + dx[d] && x + dx[d] < n && 
                0 <= y + dy[d] && y + dy[d] < n && 
                ans[x + dx[d]][y + dy[d]] == -1) {
            } else {
                d = d + 1 & 3;
            }
            x += dx[d];
            y += dy[d];
        }
        return ans;
    }
}

File: algorithms/leet.0059.src.1.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        m = n
        ans = [[0] * n for i in range(n)]
        i = 0
        j = 0
        x = 1
        for k in range(min(m, n)):
            ans[i][j] = x
            x += 1
            if k % 2:
                for p in range(m-1):
                    j -= 1
                    ans[i][j] = x
                    x += 1
                for p in range(n-1):
                    i -= 1
                    ans[i][j] = x
                    x += 1
                j += 1
            else:
                for p in range(m-1):
                    j += 1
                    ans[i][j] = x
                    x += 1
                for p in range(n-1):
                    i += 1
                    ans[i][j] = x
                    x += 1
                j -= 1
            m -= 1
            n -= 1
        return ans
